contracts/RMRK/extension/secureTokenTransferProtocol/IRMRKSecureTokenTransferProtocol.sol

```diff
@@ -0,0 +1,116 @@
+        TokenType tokenType,
+        uint256 indexed fromTokenId,
+        uint256 heldTokenId,
+        address indexed to,
+        uint256 amount
+    );
+
+    /**
+     * @notice Used to retrieve the given token's balance of given token
+     * @dev When retrieving the balance of an ERC-20 token, the `heldTokenId` parameter MUST be ignored.
+     * @param erc20Contract The address of the held token's smart contract
+     * @param tokenType The type of the token being checked for balance
+     * @param tokenId The ID of the token being checked for balance
+     * @param heldTokenId The ID of the held token of which the balance is being retrieved
+     * @return The amount of the specified ERC-20 tokens owned by a given token
+     */
+    function balanceOfToken(
+        address erc20Contract,
+        TokenType tokenType,
+        uint256 tokenId,
+        uint256 heldTokenId
+    ) external view returns (uint256);
+
+    /**
+     * @notice Transfer held tokens from a specific token.
+     * @dev The balance MUST be transferred from this smart contract.
+     * @dev Implementers should validate that the `msg.sender` is either the token owner or approved to manage it before
+     *  calling this.
+     * @dev If the token type is `ERC-20`, the `heldTokenId` MUST be ignored.
+     * @dev IF the token type is `ERC-721`, the `amount` MUST be ignored.
+     * @param tokenContract The address of the held token's smart contract
+     * @param tokenType The type of the token being transferred
+     * @param tokenId The ID of the token to transfer the held token from
+     * @param heldTokenId The ID of the held token to transfer
+     * @param amount The number of held tokens to transfer
+     * @param to The address to transfer the held tokens to
+     * @param data Additional data with no specified format, to allow for custom logic
+     */
+    function transferHeldTokenFromToken(
+        address tokenContract,
+        TokenType tokenType,
+        uint256 tokenId,
+        uint256 heldTokenId,
+        uint256 amount,
+        address to,
+        bytes memory data
+    ) external;
+
+    /**
+     * @notice Transfer tokens to a specific holder token.
+     * @dev The token smart contract must have approval for this contract to transfer the tokens.
+     * @dev The balance MUST be transferred from the `msg.sender`.
+     * @dev If the token type is `ERC-20`, the `heldTokenId` MUST be ignored.
+     * @dev If the token type is `ERC-721`, the `amount` MUST be ignored.
+     * @param tokenContract The address of the token smart contract
+     * @param tokenType The type of the token being transferred
+     * @param tokenId The ID of the token to transfer the tokens to
+     * @param heldTokenId The ID of the held token to transfer
+     * @param amount The number of ERC-20 tokens to transfer
+     * @param data Additional data with no specified format, to allow for custom logic
+     */
+    function transferHeldTokenToToken(
```

Comment:
What a wonderful name. Makes it very evident.

Reply:
It feels a bit clunky to me in some places, but I think it's clear about which token is referred to where 😅